### PR TITLE
KEYCLOAK-11912 Fix path to keytool for OpenJ9

### PR DESCRIPTION
**Pull request by hmlnarik**
Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh [hunk 1]
```diff
@@ -37,6 +37,10 @@ function autogenerate_keystores() {
     local PASSWORD=$(openssl rand -base64 32)
     local JKS_KEYSTORE_FILE="${KEYSTORE_TYPE}-keystore.jks"
     local PKCS12_KEYSTORE_FILE="${KEYSTORE_TYPE}-keystore.pk12"
+    local KEYTOOL="$(which keytool 2>/dev/null)"
+    if [ ! -x "$KEYTOOL" -a -x "$JAVA_HOME/bin/keytool" ]; then
+        KEYTOOL="$JAVA_HOME/bin/keytool"
```
Comment (iankko) [inline modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh:45]: @hmlnarik 
Wondering, should this be using the `keytool` from OpenJ9 JDK or JRE? AFAICT there seems to be two locations per:
```
# rpm -q adoptopenjdk-8-openj9
adoptopenjdk-8-openj9-8u232_b09.openj9_0.17.0-1.x86_64
```

```
# rpm -ql adoptopenjdk-8-openj9 | grep keytool
/usr/lib/jvm/adoptopenjdk-8-openj9/bin/keytool
/usr/lib/jvm/adoptopenjdk-8-openj9/jre/bin/keytool
/usr/lib/jvm/adoptopenjdk-8-openj9/man/ja/man1/keytool.1
/usr/lib/jvm/adoptopenjdk-8-openj9/man/ja_JP.UTF-8/man1/keytool.1
/usr/lib/jvm/adoptopenjdk-8-openj9/man/man1/keytool.1
```

The reason, why I am asking being, if the JDK path of the `keytool` would be used, the path provided to the `SYSTEM_CACERTS` variable below would not work. See also comment below.

File: modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh [hunk 2]
```diff
@@ -90,10 +94,10 @@ function autogenerate_keystores() {
     fi
 
     # Import existing system CA certificates into the newly generated truststore
-    local SYSTEM_CACERTS=$(readlink -e $(dirname $(readlink -e $(which keytool)))"/../lib/security/cacerts")
-    if keytool -v -list -keystore "${SYSTEM_CACERTS}" -storepass "changeit" > /dev/null; then
+    local SYSTEM_CACERTS=$(readlink -e $(dirname $(readlink -e $(which "$KEYTOOL")))"/../lib/security/cacerts")
```
Comment (iankko) [inline modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh]: @hmlnarik 

For OpenJ9 (tried on `adoptopenjdk-8-openj9-8u232_b09.openj9_0.17.0-1.x86_64`) the path to `cacerts` file seems to be:
```
# rpm -ql adoptopenjdk-8-openj9 | grep cacerts
/usr/lib/jvm/adoptopenjdk-8-openj9/jre/lib/security/cacerts
```
Notice the more `/jre`added to the path. If `keytool` from `/usr/lib/jvm/adoptopenjdk-8-openj9/jre/bin/keytool` (the JRE one` would be used, the above `SYSTEM_CACERTS` path would work. But if the JDK OpenJ9 keytool would be used (`/usr/lib/jvm/adoptopenjdk-8-openj9/bin/keytool`) the `SYSTEM_CACERTS` path argument `"/../lib/security/cacerts"` should be replaced with `"/../jre/lib/security/cacerts"` instead IMHO.

Should both OpenJ9 paths of `keytool` tool be supported? If so, the `SYSTEM_CACERTS` should append a slightly different suffix in either of those cases in that case. WDYT?

Comment (hmlnarik) [reply modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh]: I'd say we only use the `jre` one then.

Comment (iankko) [reply modules/sso/config/launch/setup/73/added/launch/openshift-x509.sh]: @hmlnarik Ok, so in that case could you update the PR so it contains `jre` in the alternative path to the `keytool` binary? Something along the lines (but untested):

```
 if [ ! -x "$KEYTOOL" -a -x "$JAVA_HOME/jre/bin/keytool" ]; then
        KEYTOOL="$JAVA_HOME/jre/bin/keytool"
```

AFAICT / AFAICS this should be sufficient, the `SYSTEM_CACERTS` to be set correctly also in the alternative case (but would be good to test & verify it of course)

Thanks!

**Comment by iankko** [comment]
@hmlnarik Tested this one a bit more. Looks `which keytool` never ends up with error, even if the OpenJ9 is the only Java installed on the system:

```
[root@localhost ~]# cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.7 (Maipo)
```
```
[root@localhost ~]# rpm -qa | grep jdk
adoptopenjdk-8-openj9-8u232_b09.openj9_0.17.0-1.x86_64
copy-jdk-configs-3.3-10.el7_5.noarch
```
```
[root@localhost ~]# rpm -qa | grep java
tzdata-java-2019c-1.el7.noarch
python-javapackages-3.4.1-11.el7.noarch
```
```
[root@localhost ~]# file $(which keytool)
/usr/bin/keytool: symbolic link to `/etc/alternatives/keytool'
```
```
[root@localhost ~]# file /etc/alternatives/keytool
/etc/alternatives/keytool: symbolic link to `/usr/lib/jvm/adoptopenjdk-8-openj9/bin/keytool'
```

looks `alternatives` is able to locate the `keytool` and chooses the JDK version of the `keytool` binary from the `adoptopenjdk-8-openj9-8u232_b09.openj9_0.17.0-1.x86_64` (or any other version RPM).

Which got me to think, if instead of adjusting the `keytool` executable path, we could just derive the root JAVA directory from it, and search for the `cacerts` file within that tree (in the directory one level up the `/bin` directory and below it).

Something like the following:
```
--- openshift-x509.sh.orig	2019-11-14 13:18:06.538000000 +0100
+++ openshift-x509.sh	2019-11-14 13:30:22.185000000 +0100
@@ -41,10 +41,6 @@ function autogenerate_keystores() {
     local PASSWORD=$(openssl rand -base64 32)
     local JKS_KEYSTORE_FILE="${KEYSTORE_TYPE}-keystore.jks"
     local PKCS12_KEYSTORE_FILE="${KEYSTORE_TYPE}-keystore.pk12"
-    local KEYTOOL="$(which keytool 2>/dev/null)"
-    if [ ! -x "$KEYTOOL" -a -x "$JAVA_HOME/bin/keytool" ]; then
-        KEYTOOL="$JAVA_HOME/jre/bin/keytool"
-    fi
 
     if [ -d "${X509_KEYSTORE_DIR}" ]; then
 
@@ -57,7 +53,7 @@ function autogenerate_keystores() {
       -out "${KEYSTORES_STORAGE}/${PKCS12_KEYSTORE_FILE}" \
       -password pass:"${PASSWORD}" >& /dev/null
 
-      "$KEYTOOL" -importkeystore -noprompt \
+      keytool -importkeystore -noprompt \
       -srcalias "${NAME}" -destalias "${NAME}" \
       -srckeystore "${KEYSTORES_STORAGE}/${PKCS12_KEYSTORE_FILE}" \
       -srcstoretype pkcs12 \
@@ -89,7 +85,7 @@ function autogenerate_keystores() {
     log_info "Creating RH-SSO truststore.."
     csplit -s -z -f crt- "${X509_CA_BUNDLE}" "${X509_CRT_DELIMITER}" '{*}'
     for CERT_FILE in crt-*; do
-      "$KEYTOOL" -import -noprompt -keystore "${JKS_TRUSTSTORE_PATH}" -file "${CERT_FILE}" \
+      keytool -import -noprompt -keystore "${JKS_TRUSTSTORE_PATH}" -file "${CERT_FILE}" \
       -storepass "${PASSWORD}" -alias "service-${CERT_FILE}" >& /dev/null
     done
 
@@ -98,10 +94,16 @@ function autogenerate_keystores() {
     fi
 
     # Import existing system CA certificates into the newly generated truststore
-    export SYSTEM_CACERTS=$(readlink -e $(dirname $(readlink -e $(which "$KEYTOOL")))"/../lib/security/cacerts")
-    if "$KEYTOOL" -v -list -keystore "${SYSTEM_CACERTS}" -storepass "changeit" > /dev/null; then
+    # Locate the system CA bundle file by searching the file system for 'cacerts' file,
+    # using directory of the "keytool" binary (without the '/bin' suffix) as the root
+    # of the directory tree to be searched
+    export SYSTEM_CACERTS=$(find $(sed -e 's#/bin$##' <<< "$(readlink -e $(dirname $(readlink -e $(which keytool))))") -name cacerts)
+    if [ -z "${SYSTEM_CACERTS}" ]; then
+      log_error "Failed to locate the default Java CA certificate bundle file on the system!"
+    fi
+    if keytool -v -list -keystore "${SYSTEM_CACERTS}" -storepass "changeit" > /dev/null; then
       log_info "Importing certificates from system's Java CA certificate bundle into RH-SSO truststore.."
-      "$KEYTOOL" -importkeystore -noprompt \
+      keytool -importkeystore -noprompt \
       -srckeystore "${SYSTEM_CACERTS}" \
       -destkeystore "${JKS_TRUSTSTORE_PATH}" \
       -srcstoretype jks -deststoretype jks \
```

seems to be working for me (even for OpenJ9):

```
[root@localhost ~]# ./openshift-x509.sh
Creating RH-SSO truststore..
RH-SSO truststore successfully created at: /tmp/opt/eap/keystores/truststore.jks
Importing certificates from system's Java CA certificate bundle into RH-SSO truststore..
Successfully imported certificates from system's Java CA certificate bundle into RH-SSO truststore at: /tmp/opt/eap/keystores/truststore.jks
SYSTEM_CACERTS was set to: /usr/lib/jvm/adoptopenjdk-8-openj9/jre/lib/security/cacerts
```

WDYT?

But if the OpenJ9 JDK / JRE is expected to be installed also other way than just from RPM, it would need possibly need to be adjusted a bit yet.

**Comment by hmlnarik** [comment]
@iankko Keytool is not found if openjdk is not installed and the command terminates with error

**Comment by iankko** [comment]
Yet, similar change should be applied to Keycloak upstream / vanilla images too (they are using the same x509 functionality too AFAICT).

**Review by iankko — APPROVED** [review]
LGTM

**Pull request merged**